OCPBUGS-33700:Added routes note for soure IP address in IController doc (#78097)
openshift/openshift-docs: modules/nw-ingress-controller-configuration-proxy-protocol.adoc

Reviewer: This is out of scope to your work and you don't have to do this at all, but if you're feeling up for it: [this line](https://github.com/openshift/openshift-docs/pull/78097/files#diff-da5babc2dce52b79d99be94bdaf4e0890583b18c2e7d9ebf8d42f1cb5f2547a3R38) (step one of this procedure) could have "by running the following command" added to the end of it.

Author: Thanks, valid. Updated. Might as well while I'm strolling through this part of the doc woods!

// Module included in the following assemblies:
//

> // * ingress/ingress-operator.adoc

Reviewer 2: Should this be `networking/ingress-operator.adoc`?

Author: Correct. Updated.

![Screenshot from 2024-06-28 11-38-22](https://github.com/openshift/openshift-docs/assets/57954076/3d0effbe-692b-4343-8858-94f5e4f56745)

:_mod-docs-content-type: PROCEDURE
[id="nw-ingress-controller-configuration-proxy-protocol_{context}"]
= Configuring the PROXY protocol for an Ingress Controller

> A cluster administrator can configure link:https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt[the PROXY protocol] when an Ingress Controller uses either the `HostNetwork` or `NodePortService` endpoint publishing strategy types.

Author: Thanks, @ShudiLi . Updated.

> A cluster administrator can configure link:https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt[the PROXY protocol] when an Ingress Controller uses either the `HostNetwork`, `NodePortService`, or 'Private' endpoint publishing strategy types.

Reviewer: These should be backticks - Private doesn't render as code text in the preview:
```suggestion
A cluster administrator can configure link:https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt[the PROXY protocol] when an Ingress Controller uses either the `HostNetwork`, `NodePortService`, or `Private` endpoint publishing strategy types.
```

Author: :heavy_check_mark:

[WARNING]
====

> The PROXY protocol is unsupported for the default Ingress Controller with installer-provisioned clusters on non-cloud platforms that use a Keepalived Ingress VIP.

Reviewer 3: Consider expanding "VIP" on first use.

Reviewer 3: Could rewrite to be active, e.g. "The default Ingress Controller with installer-provisioned clusters on non-cloud platforms that use a Keepalived Ingress VIP do not support the PROXY protocol."

Author: :heavy_check_mark:

====

> The PROXY protocol enables the load balancer to preserve the original client addresses for connections that the Ingress Controller receives. The original client addresses are useful for logging, filtering, and injecting HTTP headers. In the default configuration, the connections that the Ingress Controller receives only contain the source IP address that is associated with the load balancer.

Reviewer: Nit, per the [ISG](https://www.ibm.com/docs/en/ibm-style?topic=grammar-adverbs-only) the word "only" should be as close to the phrase it modifies:
```suggestion
The PROXY protocol enables the load balancer to preserve the original client addresses for connections that the Ingress Controller receives. The original client addresses are useful for logging, filtering, and injecting HTTP headers. In the default configuration, the connections that the Ingress Controller receives contain only the source IP address that is associated with the load balancer.
```
If you feel that the original placement is the more accurate meaning, feel free to ignore this comment

Author: :heavy_check_mark:

[IMPORTANT]
====

> For a passthrough route configuration, servers in an {product-title} cluster cannot observe the original client source IP address. If you need to know the original client source IP address, ensure that you configure Ingress access logging for your Ingress Controller, so that you can view the client source IP addresses.

Reviewer 3: ```suggestion
For a passthrough route configuration, servers in an {product-title} cluster cannot observe the original client source IP address. If you need to know the original client source IP address, configure Ingress access logging for your Ingress Controller so that you can view the client source IP addresses.
```

Author: :heavy_check_mark:

> For a passthrough route configuration, servers in an {product-title} cluster cannot observe the original client source IP address. If you need to know the original client source IP address, configure Ingress access logging for your Ingress Controller so that you can view the client source IP addresses.

Reviewer: Very unfortunate - in the ROSA doc, "an {product-title} cluster" is incorrect because it renders to "an Red Hat OpenShift...".

Up to you how to resolve this, but I feel like the simplest option would be to remove the article and make cluster plural:
```suggestion
For a passthrough route configuration, servers in {product-title} clusters cannot observe the original client source IP address. If you need to know the original client source IP address, configure Ingress access logging for your Ingress Controller so that you can view the client source IP addresses.
```

Author: Well spotted!

:heavy_check_mark:

> For re-encrypt and edge routes, the {product-title} router sets the `Forwarded` and `X-Forwarded-For` headers to enable servers to check the client source IP address.

Author: Thanks. Updated!

> For re-encrypt and edge routes, the {product-title} router sets the `Forwarded` and `X-Forwarded-For` headers to enable the application workloads to check the client source IP address.

Reviewer 3: re: "enable" in general, consider a different verb. ISG prefers to reserve that for cases where you are enabling/disabling a functionality.

Author: :heavy_check_mark:

For more information about Ingress access logging, see "Configuring Ingress access logging".
====

> Configuring the PROXY protocol for an Ingress Controller is not supported when using the `LoadBalancerService` endpoint publishing strategy type. This restriction is because when {product-title} runs in a cloud platform, and an IngressController specifies that a service load balancer should be used, the Ingress Operator configures the load balancer service and enables the PROXY protocol based on the platform requirement for preserving source addresses.

Reviewer: Missing space here?
```suggestion
Configuring the PROXY protocol for an Ingress Controller is not supported when using the `LoadBalancerService` endpoint publishing strategy type. This restriction is because when {product-title} runs in a cloud platform, and an Ingress Controller specifies that a service load balancer should be used, the Ingress Operator configures the load balancer service and enables the PROXY protocol based on the platform requirement for preserving source addresses.
```

Author: :heavy_check_mark:

[IMPORTANT]
====

> You must configure both {product-title} and the external load balancer to either use the PROXY protocol or to use TCP.

Reviewer 3: ```suggestion
You must configure both {product-title} and the external load balancer to use either the PROXY protocol or TCP.
```

Author: :heavy_check_mark:

For re-encrypt and edge routes, the {product-title} router sets the `Forwarded` and `X-Forwarded-For` headers so that application workloads check the client source IP address.

For more information about Ingress access logging, see "Configuring Ingress access logging".
====

Configuring the PROXY protocol for an Ingress Controller is not supported when using the `LoadBalancerService` endpoint publishing strategy type. This restriction is because when {product-title} runs in a cloud platform, and an Ingress Controller specifies that a service load balancer should be used, the Ingress Operator configures the load balancer service and enables the PROXY protocol based on the platform requirement for preserving source addresses.

[IMPORTANT]
====
You must configure both {product-title} and the external load balancer to use either the PROXY protocol or TCP.
====

.Prerequisites
* You created an Ingress Controller.

.Procedure

> . Edit the Ingress Controller resourc by entering the following command in your CLI:

Reviewer: One more nit: 
```suggestion
. Edit the Ingress Controller resource by entering the following command in your CLI:
```

Author: :heavy_check_mark:

[id="nw-ingress-controller-configuration-proxy-protocol_{context}"]
= Configuring the PROXY protocol for an Ingress Controller

> A cluster administrator can configure https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt[the PROXY protocol] when an Ingress Controller uses either the `HostNetwork` or `NodePortService` endpoint publishing strategy types. The PROXY protocol enables the load balancer to preserve the original client addresses for connections that the Ingress Controller receives. The original client addresses are useful for logging, filtering, and injecting HTTP headers. In the default configuration, the connections that the Ingress Controller receives only contain the source address that is associated with the load balancer.

Reviewer 2: Since [OCPBUGS-3559](https://issues.redhat.com/browse/OCPBUGS-3559) was fixed in OCP 4.12.0 and backported to OCP 4.11.40 as [OCPBUGS-33700](https://issues.redhat.com/browse/OCPBUGS-33700), PROXY protocol can also be enabled for the `Private` endpoint publishing strategy type.  Would you like me to file a separate bug report for that?

Reviewer 2: I filed [OCPBUGS-36292](https://issues.redhat.com/browse/OCPBUGS-36292) for this issue.

[IMPORTANT]
====

> For a passthrough route configuration, client pods in an {product-title} cluster cannot access their original client source IP address from within the cluster. Ensure that you configure Ingress access logging for your Ingress Controller so that you can view what source IP addresses client pods can access from within your cluster.

Reviewer 2: A few corrections:

- This issue doesn't only apply to client pods; it applies to all client connections.  
- It isn't the client that would read the source IP address; it's the server pod.  
- Using term "access" could be misunderstood to imply that the pod cannot connect to the IP address; maybe "observe" would be clearer?
- As there is some cost to recording and storing access logs, I believe it would be better to instruct the cluster-admin to enable access logs only if needed.
```suggestion
For a passthrough route configuration, servers in an {product-title} cluster cannot observe the original client source IP address. If you need to know the original client source IP address, ensure that you configure Ingress access logging for your Ingress Controller so that you can view the client source IP addresses.  
```

Author: :heavy_check_mark:

> For re-encrypt and edge routes, you can set the `X-Forwarded-For` value in the `httpHeaders` of the `ingresscontrollers.operator.openshift.io` resource to allow the client source IP address to reach client pods in a cluster.

Reviewer 2: OpenShift router sets the `forwarded` and `x-forwarded-for` headers itself; in the general case, there should be no need to configure `httpHeaders` actions to set these headers.
```suggestion
For re-encrypt and edge routes, {product-title} router sets the `Forwarded` and `X-Forwarded-For` headers to enable servers to check the client source IP address.
```

Author: :heavy_check_mark:

> For more information about Ingress access logging, see the "Configuring Ingress access logging".

Reviewer 2: Should this reference be a xref?

Author: Unfortunetly, xref-support at the module-level is not possible. Instead, it's at the assembly level (higher-level directory).

Reviewer 2: One other thing: Should the "the" be removed?
```suggestion
For more information about Ingress access logging, see "Configuring Ingress access logging".
```
Or:
```suggestion
For more information about Ingress access logging, see the "Configuring Ingress access logging" section.
```

Author: Updated

====

> Configuring the PROXY protocol for an Ingress Controller is not supported in cloud deployments. This restriction is because when {product-title} runs in a cloud platform, and an IngressController specifies that a service load balancer should be used, the Ingress Operator configures the load balancer service and enables the PROXY protocol based on the platform requirement for preserving source addresses.

Reviewer 2: There may be some confusion resulting from the fact that the _default_ endpoint publishing strategy type on cloud platforms is `LoadBalancerService`; however, the other endpoint publishing strategies can still be specified explicitly, and then it is possible to enable PROXY protocol.  So it would be more accurate to say that PROXY protocol cannot be configured _when using `LoadBalancerService`_.  
```suggestion
Configuring the PROXY protocol for an Ingress Controller is not supported when using the `LoadBalancerService` endpoint publishing strategy type. This restriction is because when {product-title} runs in a cloud platform, and an IngressController specifies that a service load balancer should be used, the Ingress Operator configures the load balancer service and enables the PROXY protocol based on the platform requirement for preserving source addresses.
```

Author: Thanks. Updated.

. Set the PROXY configuration:
+

> * If your Ingress Controller uses the `hostNetwork` endpoint publishing strategy type, set the `spec.endpointPublishingStrategy.hostNetwork.protocol` subfield to `PROXY`:

Reviewer 2: `HostNetwork` should be capitalized.
```suggestion
* If your Ingress Controller uses the `HostNetwork` endpoint publishing strategy type, set the `spec.endpointPublishingStrategy.hostNetwork.protocol` subfield to `PROXY`:
```

Author: Updated.

. Set the PROXY configuration:
+

> * If your Ingress Controller uses the `HostNetwork` endpoint publishing strategy type, set the `spec.endpointPublishingStrategy.hostNetwork.protocol` subfield to `PROXY`:

Reviewer: Based on the source block below, should the H be capitalized?
```suggestion
* If your Ingress Controller uses the `hostNetwork` endpoint publishing strategy type, set the `spec.endpointPublishingStrategy.hostNetwork.protocol` subfield to `PROXY`:
```

Author: :heavy_check_mark:

Reviewer 4: There are the two keywords in the below IC configuration file:
```
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: int1
  namespace: openshift-ingress-operator
spec:
  replicas: 1
  endpointPublishingStrategy:
    hostNetwork:
      protocol: PROXY
    type: HostNetwork
  domain: int1.shudi-4v702.qe.devcluster.openshift.com
```

Author: Thanks, @ShudiLi

        protocol: PROXY
      type: HostNetwork
----

> * If your Ingress Controller uses the `NodePortService` endpoint publishing strategy type, set the `spec.endpointPublishingStrategy.nodePort.protocol` subfield to `PROXY`:

Reviewer 4: can add the following for supporting the Private endpoint publishing strategy type:
If your Ingress Controller uses the `Private` endpoint publishing strategy type, set the `spec.endpointPublishingStrategy.private.protocol` subfield to `PROXY`:
.Sample `private` configuration to `PROXY`
spec:
  endpointPublishingStrategy:
    type: Private
    private:
      protocol: PROXY

Author: :heavy_check_mark:
